propagators/opentelemetry-propagator-instana/src/InstanaPropagator.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Context,
+  trace,
+  isSpanContextValid,
+  isValidSpanId,
+  isValidTraceId,
+  TextMapGetter,
+  TextMapPropagator,
+  TextMapSetter,
+  TraceFlags,
+} from '@opentelemetry/api';
+
+/** Instana header names */
+export const INSTANA_TRACE_ID_HEADER = 'X-INSTANA-T';
+export const INSTANA_SPAN_ID_HEADER = 'X-INSTANA-S';
+export const INSTANA_LEVEL_HEADER = 'X-INSTANA-L';
+
+const FIELDS = [
+  INSTANA_TRACE_ID_HEADER,
+  INSTANA_SPAN_ID_HEADER,
+  INSTANA_LEVEL_HEADER,
+];
+
+function readHeader(
+  carrier: unknown,
+  getter: TextMapGetter,
+  key: string
+): string {
+  let header = getter.get(carrier, key);
+  if (Array.isArray(header)) [header] = header;
+  return header || '';
+}
+
+/**
+ * Propagator for the Instana HTTP headers.
+ */
+export class InstanaPropagator implements TextMapPropagator {
+  /**
+   * Injects the current span context into Instana's vendor specific trace correlation headers (X-INSTANA-T, X-INSTANA-S
+   * and X-INSTANA-L).
+   */
+  inject(context: Context, carrier: unknown, setter: TextMapSetter) {
+    const spanContext = trace.getSpan(context)?.spanContext();
+    if (!spanContext || !isSpanContextValid(spanContext)) {
+      return;
+    }
+
+    setter.set(carrier, INSTANA_TRACE_ID_HEADER, spanContext.traceId);
+    setter.set(carrier, INSTANA_SPAN_ID_HEADER, spanContext.spanId);
+    console.log(JSON.stringify(spanContext));
+    const sampled =
+      (spanContext.traceFlags & TraceFlags.SAMPLED) == TraceFlags.SAMPLED;
+    setter.set(carrier, INSTANA_LEVEL_HEADER, sampled ? '1' : '0');
+  }
+
+  /**
+   * Extracts the span context from Instana's vendor specific trace correlation headers (X-INSTANA-T, X-INSTANA-S
+   * and X-INSTANA-L).
+   */
+  extract(context: Context, carrier: unknown, getter: TextMapGetter): Context {
+    let traceId = readHeader(carrier, getter, INSTANA_TRACE_ID_HEADER);
+    if (traceId.length < 32) {
+      traceId = traceId.padStart(32, '0');
+    }
+    const spanId = readHeader(carrier, getter, INSTANA_SPAN_ID_HEADER);
+    if (spanId.length < 16) {
+      traceId = traceId.padStart(16, '0');
+    }
+    const level = readHeader(carrier, getter, INSTANA_LEVEL_HEADER) || '1';
+    const sampled = level !== '0';
+    const traceFlags = sampled ? TraceFlags.SAMPLED : TraceFlags.NONE;
+
+    if (isValidTraceId(traceId) && isValidSpanId(spanId)) {
```

Comment:
TBD: Here we only "attach" the headers when there is a valid traceId and spanId.

But when the previous service never created a span because it is suppressed, we still want to forward this information or? Service A uses Instana, B uses Otel, C uses Instana etc.

Other than that the sampler needs access to the suppression header for any case.

Reply:
Right, good catch! I'll fix that 👍

Reply:
This is fixed now. There is a new test case for that (`extracts sampled=false from X-INSTANA-L=0, even if trace ID and span ID are missing`).